types/context.go

```diff
@@ -16,167 +13,84 @@ import (
 )
 
 /*
-The intent of Context is for it to be an immutable object that can be
-cloned and updated cheaply with WithValue() and passed forward to the
-next decorator or handler. For example,
-
- func MsgHandler(ctx Context, tx Tx) Result {
- 	...
- 	ctx = ctx.WithValue(key, value)
- 	...
- }
+Context is an immutable object contains all information needed to
+process a request.
 */
 type Context struct {
-	context.Context
-	pst *thePast
-	gen int
-	// Don't add any other fields here,
-	// it's probably not what you want to do.
-}
-
-// create a new context
-func NewContext(ms MultiStore, header abci.Header, isCheckTx bool, logger log.Logger) Context {
-	c := Context{
-		Context: context.Background(),
-		pst:     newThePast(),
-		gen:     0,
-	}
-
-	c = c.WithMultiStore(ms)
-	c = c.WithBlockHeader(header)
-	c = c.WithChainID(header.ChainID)
-	c = c.WithIsCheckTx(isCheckTx)
-	c = c.WithTxBytes(nil)
-	c = c.WithLogger(logger)
-	c = c.WithVoteInfos(nil)
-	c = c.WithGasMeter(stypes.NewInfiniteGasMeter())
-	c = c.WithMinGasPrices(DecCoins{})
-	c = c.WithConsensusParams(nil)
-	c = c.WithEventManager(NewEventManager())
-
-	return c
-}
-
-// is context nil
-func (c Context) IsZero() bool {
-	return c.Context == nil
-}
-
-// ----------------------------------------------------------------------------
-// Getters
-// ----------------------------------------------------------------------------
-
-type contextKey int // local to the context module
-
-const (
-	contextKeyMultiStore contextKey = iota
-	contextKeyBlockHeader
-	contextKeyChainID
-	contextKeyIsCheckTx
-	contextKeyTxBytes
-	contextKeyLogger
-	contextKeyVoteInfos
-	contextKeyGasMeter
-	contextKeyBlockGasMeter
-	contextKeyMinGasPrices
-	contextKeyConsensusParams
-	contextKeyEventManager
-)
-
-// context value for the provided key
-func (c Context) Value(key interface{}) interface{} {
-	value := c.Context.Value(key)
-	if cloner, ok := value.(cloner); ok {
-		return cloner.Clone()
-	}
-	if message, ok := value.(proto.Message); ok {
-		return proto.Clone(message)
-	}
-	return value
+	ctx           context.Context
```

Comment:
I don't see the advantage of having `Context` no longer extend `context.Context` - I would prefer to see this change reversed

Reply:
I agree with @rigelrozanski here

Reply:
I don't see any advantages there. I will discuss usage of Context elsewhere, but technically

```go
type Context interface {
	Deadline() (deadline time.Time, ok bool)
	Done() <-chan struct{}
	Err() error
	Value(key interface{}) interface{}
}
```

Only `Value` is explicitly implemented by `sdk.Context`. If we try to use it as a normal Context object, with any of the standard functions, we quickly get issues. (This code was checked against the master branch):

```go
func demo() {
	var ctx Context
	ctx, _ = context.WithCancel(ctx)
}
```

-> `types/context.go:64:9: cannot assign context.Context to ctx (type Context) in multiple assignment: need type assertion`

So we pretend to be a `context.Context`, but if we pass it in like that, we are no longer `sdk.Context`. Which means embedding it this was is actually *dangerous*.

By including Context as a member, we can then properly do the above desired issue (this code off of the PR):

```go
func demo() {
	var ctx Context
	std, cancel := context.WithTimeout(ctx.Context(), time.Second)
	defer cancel()
	ctx = ctx.WithContext(std)
	select {
	case <-ctx.Context().Done():
			fmt.Println("Done!")
	}
}
```

Note that the compiler does not let me try to pass in `sdk.Context` as `context.Context`, which is a benefit.

Reply:
I also think sdk.Context is a misnomer. 
It is really more like a Request.

I prefer Context for holding non-necessary values, not critical points like the DB itself. (not to say other frameworks, like gin, don't use this naming of Context as well)

Reply:
I think I've changed my mind.

> So we pretend to be a context.Context, but if we pass it in like that, we are no longer sdk.Context. Which means embedding it this was is actually dangerous.

I always thought that Golang's context was a pretty clunky solution to two problems:
1. HTTP request-specific values
2. cancellation

IMvHO Golang does not elegantly solve either. Plus, our custom Context was not even designed to address 2. and Golang 2 will supposedly solve 2. once and for all.

In conclusion:

* ACK on `Context -> Request` renaming
* ACK on the removal of the composition with `context.Context`

Reply:
> I always thought that Golang's context was a pretty clunky solution to two problems:
>   1. HTTP request-specific values
>   2. cancellation

I agree with you here. It does work as a generic store (1), but that is usually a sign of a design flaw. I mean a `map` or `array` with a specific type could often be a better solution (as you know in advance which way you want to extend). That said, I don't think we need to remove that functionality. I would rather deprecate it until we ensure there are no needs for it.

The only real bonus I find is that some frameworks like [monkit](https://github.com/spacemonkeygo/monkit) will "auto"-instrument your code if you pass them Context. I see main value just for interoperating with other libs.

Reply:
@rigelrozanski if the issue is verbose-ness (which is true), I am happy to rename the methods to `Ctx()` and `WithCtx()`, which make it shorter, but possibly less clear read.

Reply:
Sorry I couldn't follow up on this thread took some vacation! Sounds fine though